FINERACT-2311: Buy Down Fees  - External Business Events

somasorosdpc opened the pull request:
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests
- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.
- [ ] Create/update unit or integration tests for verifying the changes made.
- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.
- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes
- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

somasorosdpc committed:
FINERACT-2311: Buy Down fees - External business events

magyari-adam commented on fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/BuyDownFeeWritePlatformServiceImpl.java:
```diff
@@ -122,7 +125,7 @@ public CommandProcessingResult makeLoanBuyDownFee(final Long loanId, final JsonC
         // Notify business events
-        businessEventNotifierService.notifyPostBusinessEvent(new LoanBuyDownFeeTransactionCreatedBusinessEvent(buyDownFeeTransaction));
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanBuyDownFeeTransactionBusinessEvent(buyDownFeeTransaction));
         businessEventNotifierService.notifyPostBusinessEvent(new LoanBalanceChangedBusinessEvent(loan));
```
Should we fire a LoanBalanceChanged event here? Buy down fees should not have any impact on the schedule.

magyari-adam commented on fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/BuyDownFeeWritePlatformServiceImpl.java:
```diff
@@ -202,14 +205,48 @@ public CommandProcessingResult buyDownFeeAdjustment(final Long loanId, final Lon
         businessEventNotifierService
-                .notifyPostBusinessEvent(new LoanBuyDownFeeAdjustmentTransactionCreatedBusinessEvent(savedBuyDownFeeAdjustment));
+                .notifyPostBusinessEvent(new LoanBuyDownFeeAdjustmentTransactionBusinessEvent(savedBuyDownFeeAdjustment));
         businessEventNotifierService.notifyPostBusinessEvent(new LoanBalanceChangedBusinessEvent(loan));
```
Should we fire a LoanBalanceChanged event here? Buy down fees should not have any impact on the schedule.

magyari-adam commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanBuyDownFeeTest.java:
```diff
@@ -53,7 +59,7 @@
  */
 @Slf4j
-@ExtendWith({ LoanTestLifecycleExtension.class })
+@ExtendWith({ ExternalEventsExtension.class })
```
BaseLoanIntegrationTest is already extended with ExternalEventsExtension.

magyari-adam committed:
FINERACT-2311: Buy Down fees - External business events

- buy down fee reversal fix
- added more listeners for buy down fee amortization on closed loans

peter-kovacs-dpc committed:
FINERACT-2311: Buy Down fees - External business events - E2E tests

adamsaghy reviewed (approved):
LGTM
